comp/otelcol/otlp/components/exporter/datadogexporter/factory.go

```
@@ -35,29 +37,46 @@ import (
 )
 
 type factory struct {
-	onceAttributesTranslator sync.Once
-	attributesTranslator     *attributes.Translator
-	attributesErr            error
-
-	registry  *featuregate.Registry
-	s         serializer.MetricSerializer
-	logsAgent logsagentpipeline.Component
-	h         serializerexporter.SourceProviderFunc
+	attributesErr          error
+	onceSetupTraceAgentCmp sync.Once
+
+	registry      *featuregate.Registry
+	s             serializer.MetricSerializer
+	logsAgent     logsagentpipeline.Component
+	h             serializerexporter.SourceProviderFunc
+	traceagentcmp traceagent.Component
 }
 
-func (f *factory) AttributesTranslator(set component.TelemetrySettings) (*attributes.Translator, error) {
-	f.onceAttributesTranslator.Do(func() {
-		f.attributesTranslator, f.attributesErr = attributes.NewTranslator(set)
+// setupTraceAgentCmp sets up the trace agent component.
+// It is needed in trace exporter to send trace and in metrics exporter to send apm stats.
+// The set up happens only once, subsequent calls are no-op.
+func (f *factory) setupTraceAgentCmp(set component.TelemetrySettings) error {
+	f.onceSetupTraceAgentCmp.Do(func() {
+		var attributesTranslator *attributes.Translator
+		attributesTranslator, f.attributesErr = attributes.NewTranslator(set)
+		if f.attributesErr != nil {
+			return
+		}
```

U need to propagate the error. One way to do named return of error `err`